mesonbuild/modules/gnome.py

```diff
@@ -1689,5 +1689,93 @@ def generate_vapi(self, state, args, kwargs):
+    def valadoc(self, state, args, kwargs):
+        build_dir = os.path.join(state.environment.get_build_dir(), state.subdir)
+        source_dir = os.path.join(state.environment.get_source_dir(), state.subdir)
+
+        if not args:
+            raise MesonException('valadoc takes at least one positional argument')
+
+        # Build the command line
+        cmd = [ self.interpreter.find_program_impl('valadoc') ]
+
+        # Always force, or it will complain that "File already exists" when
+        # doing an incremental update
+        cmd += [ '--force' ]
+
+        # Package
+        pkg_name = kwargs.get('package_name')
+        pkg_version = kwargs.get('package_version')
+        if not pkg_name:
+            raise MesonException('Missing required "package_name" keyword argument')
+        if not pkg_version:
+            raise MesonException('Missing required "package_version" keyword argument')
+        cmd += [ '--package-name=' + pkg_name ]
+        cmd += [ '--package-version=' + pkg_version ]
+        package = '{}-{}'.format(pkg_name, pkg_version)
+
+        # VAPI packages
+        cmd += self._vapi_args_to_command('--pkg=', 'packages', kwargs)
+        cmd += self._vapi_args_to_command('--vapidir=', 'vapi_dirs', kwargs)
+
+        # Doclet
+        doclet = kwargs.get('doclet')
+        if not doclet:
+            doclet = 'html'
+        cmd += [ '--doclet=' + doclet ]
+
+        doclet_args = kwargs.get('doclet_args')
+        if doclet_args is not None:
+            for d_arg in doclet_args:
+                # Prevent people from doing something funky
+                if not isinstance(d_args, str):
+                    raise MesonException('Keyword argument "doclet_args" allows only string options')
+                cmd += [ d_arg ]
+
+        # Install directory
+        install = kwargs.get('install')
+        if install:
+            install_dir = kwargs.get('install_dir')
+            if not install_dir:
+                # By default, install into ${datadir}/doc/${package}
+                datadir = state.environment.get_datadir()
+                install_dir = os.path.join(datadir, 'doc')
+
+                # For some doclet types, we can do better
+                if doclet == 'gtkdoc':
+                  install_dir = os.path.join(datadir, 'gtk-doc', 'html')
+                elif doclet == 'devhelp':
+                  install_dir = os.path.join(datadir, 'devhelp', 'books')
+
+        # Output directory
+        cmd += [ '--directory=' + os.path.join(build_dir, package) ]
+
+        # Inputs
+        inputs = self.interpreter.source_strings_to_files(args)
+        for f in inputs:
+            if isinstance(f, mesonlib.File):
+                cmd += [ f.absolute_path(state.environment.source_dir,
+                                         state.environment.build_dir) ]
+            else:
+                raise InterpreterException('valadoc inputs can only be strings or file objects')
```

Comment:
I think not support CustomTarget, CustomTargetIndex, and ConfigureFiles here is going to be a limitation that needs to be fixed at some point so we should probably do it now.

Aside -- we've reimplemented the "handle strings, files, CustomTargets, CustomTargetIndexes, and ConfigureFiles" enough times we should probably have a helper somewhere for it